packages/ckeditor5-table/src/converters/table-cell-refresh-post-fixer.js

```diff
@@ -25,79 +25,89 @@ export default function injectTableCellRefreshPostFixer( model ) {
-	let insertCount = 0;
+	const changesForCells = new Map();
+	const changes = [ ...differ.getChanges() ];
 
-	for ( const change of differ.getChanges() ) {
-		const parent = change.type == 'insert' || change.type == 'remove' ? change.position.parent : change.range.start.parent;
+	// Updated refresh algorithm.
+	// 1. Gather all changes inside table cell.
+	changes.forEach( change => {
+		const parent = change.type == 'attribute' ? change.range.start.parent : change.position.parent;
 
 		if ( !parent.is( 'element', 'tableCell' ) ) {
-			continue;
+			return;
 		}
 
-		if ( change.type == 'insert' ) {
-			insertCount++;
+		if ( !changesForCells.has( parent ) ) {
+			changesForCells.set( parent, [] );
 		}
 
-		if ( checkRefresh( parent, change.type, insertCount ) ) {
-			cellsToRefresh.add( parent );
+		changesForCells.get( parent ).push( change );
+	} );
+
+	// Stores cells to be refreshed, so the table cell will be refreshed once for multiple changes.
+	const cellsToRefresh = new Set();
+
+	// 2. For each table cell:
+	for ( const [ tableCell, changes ] of changesForCells.entries() ) {
+		// 2a. Count inserts/removes as diff and marks any attribute change.
+		const { childDiff, attribute } = getChangesSummary( changes );
+
+		// 2b. If we detect that number of children has changed...
+		if ( childDiff !== 0 ) {
+			const prevChildren = tableCell.childCount - childDiff;
+			const currentChildren = tableCell.childCount;
+
+			// Might need refresh if previous children was different from 1. Eg.: it was 2 before, now is 1.
+			if ( currentChildren === 1 && prevChildren !== 1 ) {
+				cellsToRefresh.add( tableCell );
+			}
+
+			// Might need refresh if previous children was 1. Eg.: it was 1 before, now is 5.
+			if ( currentChildren !== 1 && prevChildren === 1 ) {
+				cellsToRefresh.add( tableCell );
+			}
+		}
+
+		// ... 2c or some attribute has changed.
+		if ( attribute ) {
+			cellsToRefresh.add( tableCell );
 		}
 	}
 
+	// Having cells to refresh we need to
 	if ( cellsToRefresh.size ) {
-		// @if CK_DEBUG_TABLE // console.log( `Post-fixing table: refreshing cells (${ cellsToRefresh.size }).` );
+		// @if CK_DEBUG_TABLE // console.log( `Post-fixing table: Checking table cell to refresh (${ cellsToRefresh.size }).` );
+		// @if CK_DEBUG_TABLE // let paragraphsRefreshed = 0;
 
 		for ( const tableCell of cellsToRefresh.values() ) {
-			differ.refreshItem( tableCell );
+			for ( const paragraph of [ ...tableCell.getChildren() ].filter( child => child.is( 'element', 'paragraph' ) ) ) {
+				// @if CK_DEBUG_TABLE // console.log( `Post-fixing table: refreshing paragraph in table cell (${++paragraphsRefreshed}).` );
+				differ.refreshItem( paragraph );
+			}
 		}
-
-		return true;
 	}
 
+	// Always return false to prevent the refresh post-fixer from re-running on the same set of changes and going into an infinite loop.
+	// See https://github.com/ckeditor/ckeditor5/issues/1936.
```

Comment:
Maybe we should check if the first paragraph is not marked for refresh instead of returning false?

Reply:
Maybe this will be OK, thanks! I've got lost here TBH.

Reply:
Still, there's something odd :/ Look at the current implementation - it hangs other tests.

I'm leaving this as-is for now as I don't see investing more time into this at this stage makes sense.